### DISCUSSION TITLE
Project Management - Settings Management Clean-Up

### DISCUSSION THREAD

**Discussion by scottfurry** [sqlitebrowser/sqlitebrowser#3317] [Ideas]
### Topic Rationale
My world currently is a bit messy at the moment. I need to peel off circumstances discovered during investigations associated with an issue. Discussion topic is being used both to ensure I do not lose visibility of what was found triaging identified issue, as well as address other items that exceed scope of issue. I am initially including the issue reporter, @bartek990, in this conversation.

### Topic Background
Issue #3312 was created for awareness about problems with DB4S's response to user selected toolbar icon settings. My review of the matter determined that the setting was being established in storage. However, DB4S code may not be making use of this setting value on application start-up. There is some reporting this behavior is intermittent.

However, review of code for DB4S settings and main window appear to show larger problems with how settings are being stored, retrieved, and utilized in DB4S. Specifically, only a bare settings file is established when exiting initial usage of DB4S. Also, some settings values appear to be employed only on the `save` action of the settings dialog. Finally, I have concerns about setting key/value pairs and how the default values are being documented and/or are scattered across the DB4S code.

<s>A [comment to #3341](https://github.com/sqlitebrowser/sqlitebrowser/issues/3341#issuecomment-1541090992) identified usage of `sourceforge` domain for storage of settings. Develop code to detect this circumstance. Effort should be made to remove `sourceforge` reference and utilize current DB4S domain.</s>
_see commentary below_

Pull request #3348 identified issue where a default font is not established for databrowser control on application start-up.

### Topic Goal
Ultimately, a clear and consistent settings practice and handling are ideal. Efforts under this topic will work towards incorporating these desired changes.

### Steps
- identify all get/set setting locations (may be indicated with `Settings` namespace)
- develop and document setting keys, as well as default and allowable values for that key
  This includes platform specific defaults (#3403 identified MacOS does not have a font named `Monospace`)
- rationalize a process for how settings should be handled by DB4S
- produced a plan of what code needs to change - this will incorporate process identified above
~- develop a wizard to identify storage with `sourceforge` domain. User to be given a choice: a) leave existing settings and use; b) leave existing settings and start fresh; c) move existing settings to new location; or d) remove existing settings and start fresh.~ _see commentary below_
- complete code changes/test/push branch of all changes to main repository

I will keep others apprised of efforts on this front as my time permits.

**Comment by chrisjlocke** [comment]
> develop and document setting keys, as well as default and allowable values for that key

I'm happy to help here document and slap up a Wiki page (maybe we need a developers wiki area too?) and help trawl through code.

**Reply by justinclift** [reply]
Just to point out, we do have both "Users" and "Developers" section/topics on our wiki front page already. :smile:

**Comment by scottfurry** [comment]
[Settings_Usage.ods](https://github.com/sqlitebrowser/sqlitebrowser/files/11259501/Settings_Usage.ods)
[Usage.txt](https://github.com/sqlitebrowser/sqlitebrowser/files/11259502/Usage.txt)

Completion of first step.

From the `Usage.txt` file, the steps to reproduce the file is...
Steps:
- ensure you have latest project sources downloaded
- on command line, cd to project source directory
- execute ack (beyondgrep.org) search of cpp type sources (h,hpp,hxx,c,cpp,cxx,etc)
- command line: ack '[\s\(!]Settings::' -t cpp >[path]/Usage.txt
- manually sort with order: Settings|Application|MainWindow|Preferences|remainder case-insensitive alphabetical order

I'll spare everyone my immediate observations but there are some serious messiness.
Next up is to take this spreadsheet (produced in libreoffice Calc) to update what/where default values are established.

Work continues...

**Comment by justinclift** [comment]
Cool, this looks like a very organised approach.  Should produce a good result. :smile:

And yeah, from that Usage.txt, it does seem like an er... non-trivial amount of work.  Hopefully after you get the initial pieces into logical shape, the rest fall into line easily without too much hairiness. :smile:

**Reply by scottfurry** [reply]
Scroll the spread sheet and view some of the sections and key names. Inconsistencies abound. My head is already going to a "transitional" or "migration" strategy for previous versions. Hurts to be "big brained" some days.

**Reply by chrisjlocke** [reply]
If it was me (and its not) we would have two classes.  Class 1 would handle all the settings we need in the new places.  I would argue that 'sqlitebrowser/sqlitebrowser' in the registry was even wrong, but as that name is everywhere, including the website, that seems changing for changes sake.
Class 1 would then ask class 2 which would handle legacy settings in ye olde places.  This means when we release v4, or v5, the old v3 code (the old class) can easily be removed.
pseudo code;
program: "Can I have the setting for X?"
class 1: "Sure.  Hmm. I don't know or can't find that. Class 2?
class 2: "Here it is."
class 1: "Ta.  Here you go Mr Program. Actuallyy, that X position is out of range- must be for an older monitor - i'll do some validation before blindingly giving you a value"
program: "Awfully nice of you. What a nice chap."

This way as well, it wouldn't be such a mammoth task - the code could slowly be changed to point to the new class over a series of nightlies - it wouldn't need to be done in one mammoth update for fear of breaking anything - the two 'systems' (old broken code and new settings class) could live harmoniously.

**Reply by scottfurry** [reply]
Or we suck it up and just rip the blood soaked band-aid off the gaping wound...

**Reply by justinclift** [reply]
@scottfurry Well, you're the point person on this, so you get to pick... :smile:

... and changing your mind 1/2 way through as things proceed is allowed of course.  That's just how stuff goes sometimes. :wink:

**Reply by scottfurry** [reply]
No pressure...
I'm partial to just have a firm demarcation between settings versions. Some of what @chrisjlocke is pointing out has merit. I'm reluctant to drag out implementation. I think it would be useful to look at users and say `Version 1x.0 has this...`. We can have a shim to migrate settings from a previous version (think a GUI-based wizard function on first start). This seems the most prudent going forward in the grand/overall scheme of things.

**Reply by chrisjlocke** [reply]
One other idea that will never see the light of day is a settings database. So DB4S uses sqlite, so why not create a sqlite settings.db in the same directory as the application. Simply a key/value pair, but allows users to have different settings relatively easily - just a command-line switch to point to a different db. Simple to share settings - in fact, you could even share on DBHub.io and have your settings 'portable' if you travel a lot.
Hmm. Maybe I should fork it and see if it has traction...
I always wanted the project file to be in a database (a database for the database .. how deep can we go?) and as that's broken too at the moment, could cover two bases. (as its the same key/value class...)

**Reply by scottfurry** [reply]
Which has merit. Although INI-format leverages existing API usage from Qt for the desktop application. Key/Value database creeps into NoSQL territory.

**Reply by chrisjlocke** [reply]
> Although INI-format leverages existing API usage

Thats not a Linux friendly format .. wouldn't it be .toml or something?  Have no idea what Macs use .. whatever a .plist is?
My thought with a sqlite database is that DB4S is already a sqlite 'consumer', so using a sqlite database in a sqlite database application seems pretty trivial.

**Reply by scottfurry** [reply]
And this is what I have been mulling over as I dig out all the locations with default values....

When DB4S is first started, only a minimal number of settings are populated. On exit, these minimal settings are stored. It is only after activation of the settings dialog is fulsome, complete, settings populated/saved. I commented as such in the original bug report.

The original bug report is derived from the setting for icons not being available or persistent. The code is littered with enormous messiness for settings management. This can be cleaned up.

My thoughts currently was to define **complete** `default` settings. DB4S is started and loads the defaults from a file (JSON, TOML, etc to be determined). DB4S would then go look for user settings. Values in memory would be updated with user choices.

This mechanism would ensure that if app behavior depends on a setting, it is defined in the `default`. If user changes it, those values would be reflected in stored settings. Either way, the complete settings are always used, rather than a minimal number until the settings dialog is activated.

As for file format, INI format is currently used because it is the default/supported format within Qt. JSON has it's limitations but is most functional. We already make use of the JSON library within DB4S - extending to settings is possible. If we switch to TOML, we would require importing another library into the DB4S code base, sub-optimal situation unless we have other TOML usage.

I am incredibly reluctant to use a SQLite DB for settings. It would be a bit awkward to work with, IMHO. `default` settings would then be some opaque binary blob that would have to undergo active revisioning(i.e. DB blob is revision XXXX which corresponds to state Q of settings) - versus - this is a text file that git can show changes with time.

**Reply by justinclift** [reply]
@chrisjlocke We use TOML for the DBHub.io related configuration files as it works pretty well and is fairly popular in Go circles for that purpose.

Lots of other projects use JSON files instead, as that's the preference of their authors.

Some places do use .ini files, it's not super common, but it does happen. :wink:

It doesn't really matter, as they all work. :smile:

**Reply by justinclift** [reply]
> My thoughts currently was to define complete default settings. ...

@scottfurry No worries.  This does sound like a solid, comprehensive approach.  It'll probably take a while, but the end result sounds like it'll be very good. :smile:

**Comment by chrisjlocke** [comment]
Throwing this in here more of an observational point... it sort of belongs here but not entirely.
In this post: https://github.com/sqlitebrowser/sqlitebrowser/issues/3341#issuecomment-1541090992
The settings file had Sourceforge in the name?  Is this still relevant?  Shouldn't it be just our main domain by now?  Is this a major thing to change?

**Reply by justinclift** [reply]
Yeah, it's a leftover from when the application used to be hosted on SourceForge, years ago.

When we went to update it, it turned out people's preferences were stored under that key.  So if we change it, people will lose all their saved preferences.

Thus us leaving it alone. :smile:

**Reply by chrisjlocke** [reply]
Well, if the whole thing is getting an overhaul, the application can read in from old point A, old point B, old point C, but write to new point D.  Mind you, everything else is 'sqlitebrowser' which isn't even our name, so we may as well keep it .. why move to something equally as broken! 🤣🤣

**Reply by scottfurry** [reply]
> "Well, it's unlikely, but could you try deleting the DB4S configuration file and trying again? smile
mv $HOME/Library/Preferences/net.sourceforge.sqlitebrowser.plist $HOME/Library/Preferences/net.sourceforge.sqlitebrowser.plist.bak"

Yup, this is a holder-over from archaic times. Various libraries, include macOS, stipulate that a reverse-url should be used to identify storage. The thinking was that since a domain is unique, its reverse should also be unique. With uniqueness, storage of one applications settings should not conflict with other applications.

This can be easily fixed. Currently, DB4S has its own domain and we should use that. A wizard-like shim should be developed. User is asked, "Do you want to import old settings?" or "Start fresh from scratch?". Wizard would either move the old settings to a new location or remove the old settings entirely.

Hunting down usage of `sourceforge` and implementing is not a difficult matter. It should be done.

**Reply by justinclift** [reply]
> Hunting down usage of sourceforge and implementing is not a difficult matter. It should be done.

Heh Heh Heh.  Well, if you're up for it, why not? :smile:

**Comment by scottfurry** [comment]
[Settings_Usage.ods](https://github.com/sqlitebrowser/sqlitebrowser/files/11861477/Settings_Usage.ods)
[Usage_GetValue.txt](https://github.com/sqlitebrowser/sqlitebrowser/files/11861427/Usage_GetValue.txt)
[Usage_SetValue.txt](https://github.com/sqlitebrowser/sqlitebrowser/files/11861428/Usage_SetValue.txt)

Second Stage:
Find all section/key/value tuples and document default values

Steps to regenerate are documented in files Usage_GetValue.txt and Usage_SetValue.txt

Some Quick Observations:
- Section `db` Key `lastlocation` - inconsistent naming (not camelCase) and inconsistent usage with both singular and plural (`lastlocations`). Needs cleanup.
- Section `proxy` Key `password` - **CONFIRM PASSWORD IS NOT BEING STORED IN CLEAR!!!!**
- Section `remote` Key `clonedirectory` - inconsistent naming (not camelCase)
- I need to review Scintilla docs. Syntax colouring (section `syntaxhighlighter`) may be "customizable" (similar to how Geany does things). Future consideration...

**Comment by scottfurry** [comment]
[DB4S_Set_Cleanup.diff.zip](https://github.com/sqlitebrowser/sqlitebrowser/files/11928993/DB4S_Set_Cleanup.diff.zip)
[Usage.txt](https://github.com/sqlitebrowser/sqlitebrowser/files/11928994/Usage.txt)
[Usage_GetValue.txt](https://github.com/sqlitebrowser/sqlitebrowser/files/11928995/Usage_GetValue.txt)
[Usage_SetValue.txt](https://github.com/sqlitebrowser/sqlitebrowser/files/11928996/Usage_SetValue.txt)

First swipe at Third Stage - change.

DB4S_Set_Cleanup.diff.zip is a zip file of diff from current HEAD. I have replaced all the hard coded strings with constants in a namespace in `src/Settings.h`. The default settings are loaded in `src/Settings.cpp` into an unordered_map. There was a trick I found that allows to treat the Section name and Key name as separate entities, rather than the current practice in the settings cache of mashing the two strings together.

The original bug of not changing icon style despite user settings is still present. However, a couple of the poorly regulated hard-coded strings (see `lastlocation` description above) were cleaned up. I am not totally convinced this is the correct path to follow despite the problems cleaned up.

A difficulty I am finding is a conflation of concerns. User settings are twisted around app presentation/colour settings with SQL window Scintilla syntax highlighting. My personal preferences is to separate these things out. Application Styling in its own file(s) as part of the application install. Syntax Highlighting would be in another file(s) also installed. User *could* import other styles as desired. This path is a large bit of scope creep, IMO.

However, I want to believe things are going in an *okay* direction. I needed to do some debugging/tracing to get a better appreciation of what is happening *under the hood* when it comes to user settings.

Suggestions welcomed.

**Reply by chrisjlocke** [reply]
> Suggestions welcomed

Release v1 of the above fixes  to resolve the current setting problems / bugs.  
The separation of application/user settings can be v2.  As we don't have that functionality yet, and v1 fixes issues (which admittedly a registry file could resolve .. but not pretty), it doesn't seem right to delay v1 to implement v2.

> original bug of not changing icon style despite user settings is still present

I assume this is easier to spot and rectify once a 'central' settings strategy has been defined?  Is it clear /why/ this is still present?

**Reply by scottfurry** [reply]
The setting value is used when preferences are updated/changed. My suspicion is that the setting is not pulled in when the MainWindow is established on start-up. Tracing/Debugging should flush out where the change needs to be made.

**Comment by scottfurry** [comment]
From #3403 - MacOS does not have a font named `Monospace`. Recent build on MacOS w/ ALL_WARNINGS enabled produced the output:

```
"/Users/[me]/DB4S-src/build/DB Browser for SQLite.app/Contents/MacOS/DB Browser for SQLite"
qt.qpa.fonts: Populating font family aliases took 263 ms. Replace uses of missing font family "Monospace" with one that exists to avoid this cost. 

```

:. A default, platform-specific font will need to be defined in settings to avoid. Item added to overall list of TODO items.

**Comment by scottfurry** [comment]
And this where we dive into the weeds...

At `src/Settings.cpp:28` is the function `isValidSettingsFile`. The function signature declares a `const QString&` named `userSettingsFile` as a parameter. Problem is that the name **_shadows_** the class variable `Settings:userSettingsFile`.

A quick run through in a debugger leads me to believe I can make the shadow problem go away - simple rename. However, looking at the function and calling code gives me pause.

There is usage of `userSettingsFile != nullptr`. That seems _just wrong_. Qt _should_ establish a QString and therefore checking that the variable points to an object in memory seems wasteful. This pattern is done in the function, as well as, shortly after the return at `src/Settings.cpp:64`

From context, and I could be wrong here, the intent of the code is checking for whether the variable is populated with a usable value for the path to the settings file. Simply checking for `QString->isEmpty()` may be insufficient. I'm wondering if we should be _borrowing_ from python by changing the variable type to QPath and checking that contents reference an existing file.

Feel free to chime in here folks.

**Reply by chrisjlocke** [reply]
Upon a fresh installation, I assume userSettingsFile is null as its not set .. not until a file has been created or the user specifies one at startup?  Do we therefore expect the 'is this a valid name to a file' to legitimately fail?

**Reply by scottfurry** [reply]
_assume_ - such a dangerous word in this day and age...
If I understand things correctly, compiler will setup the memory for a `QString`, albeit uninitialized. The test `QString != nullptr` does weird things in debugging. `lldb` just skips over this and does not execute code afterwards. I have to work through a usage issue on my end. (_edited_)

However, if we test for `!QString->isEmpty()` and `QFileInfo(QString)->exists()` then the test _should_ be executed, _in theory_. I also believe that intent is more clear and robust. I'll also do away with the variable shadowing that is happening, as well.

**Comment by scottfurry** [comment]
~Settings_Clean-Up_Stage_3.diff.zip~ removed

Taking another swipe at changes. Code "looks" better. I can rationalize some of what is being done here. A key issue was the shadowing I mentioned earlier - fixed.

If executed from the console, DB4S w/ diff applied spits out when a default value is _not_ found. There are a few `syntaxhighlighter` keys that get missed. I have to debug where these values are called.

The problem with a lack of persistence of some settings keys is still present. Resolving will be another debug session.

Feel free to point out issues/problems.

**Reply by justinclift** [reply]
> If executed from the console, DB4S w/ diff applied spits out when a default value is not found.

That's a really smart idea. :smile:

**Comment by scottfurry** [comment]
_One small step for man...One giant leap forward for me..._ \o/

There's a bunch of debug statements being thrown to the console during execution. That can be trimmed down.

However, I think I have a handle on the default fixed font setting. This should silence the noise over usage of "Monospace" font on MacOS. Feel free to take the patch for a test drive.
 
~Settings_Clean-Up_Working.diff.zip~ removed

**Comment by scottfurry** [comment]
At some point, we're going to have to have a discussion about font usage.

I was able to clean-up the warning message in MacOS about missing `Monospaced` font when the app starts. Unfortunately, I opened a can of code worms here.

The application font is determined from the OS settings for a serif font. Then there is the font for the Databrowser, the SQL Editor, the Log, and <others, I'm sure>. Databrowser font is selectable in the preferences dialog, Same also for the Editor. Log, Databrowser, and SQL Editor are set to fallback to the OS defined fixed width font.

Future consideration should be given to simplifying font usage/selection. The general tab could be adjusted to have selection for a serif font and fixed width font. Other tabs with font selection would either use the application defined fixed or serif font.

At least, that's my thinking. Everyone is welcomed to chime in here...

**Comment by scottfurry** [comment]
Yeah, incremental progress...

The original bug, #3312, that started me down this path pointed out an issue with toolbar preferences. And speaking of _can of code worms_ (see my previous post about fonts)...

Looks like I'm going to have to enforce some kind of order in UI elements. I was so hoping I wouldn't have to touch these...

```diff
// MainWindow.cpp:2340
    }

-    setToolButtonStyle(static_cast<Qt::ToolButtonStyle>(Settings::getValue("General", "toolbarStyle").toInt()));
-    ui->dbToolbar->setToolButtonStyle(static_cast<Qt::ToolButtonStyle>(Settings::getValue("General", "toolbarStyleStructure").toInt()));
-    ui->toolbarSql->setToolButtonStyle(static_cast<Qt::ToolButtonStyle>(Settings::getValue("General", "toolbarStyleSql").toInt()));
+    ui->toolbarDB->setToolButtonStyle(
+            static_cast<Qt::ToolButtonStyle>(Settings::getValue(szINI::SEC_GENERAL, szINI::KEY_TB_STYLE_APP).toInt()));
+    ui->dbToolbar->setToolButtonStyle(
+            static_cast<Qt::ToolButtonStyle>(Settings::getValue(szINI::SEC_GENERAL, szINI::KEY_TB_STYLE_STRUCTURE).toInt()));
+    ui->toolbarSql->setToolButtonStyle(
+            static_cast<Qt::ToolButtonStyle>(Settings::getValue(szINI::SEC_GENERAL, szINI::KEY_TB_STYLE_SQL).toInt()));

    // Set prefetch sizes for lazy population of table models
    for(int i=0;i<ui->tabSqlAreas->count();++i)
```
There's an **ASSUMPTION** being made in the original code that calling `setToolButtonStyle` will set the Main Window ToolBar style - it should. On start up, however, the setting is ignored. Opening the preferences dialog, changing values, and saving does a whole host of things - including changing the toolbar button style - sort of. That change is not persistent on a restart.

However, if I remove the _assumption_ and explicitly reference the ui element the preference is persistent on restart. Again, incremental progress...

Unfortunately, the ExtraDB and Project toolbar button styles do not appear to be following what is in the preferences. Like I said earlier, can of code worms.

**Reply by justinclift** [reply]
Heh Heh Heh.

It's probably a really, really good thing that we have someone who knows the value of persistence working on this. :grin:

**Reply by justinclift** [reply]
Btw, not really commenting technically here because everything you're saying is making sense to me.  So, no objections or insights on my part.

**Reply by scottfurry** [reply]
Something to be said for having a modicum of obstinance. It's a coin toss between grey-hair induced, pragmatic wisdom and a healthy dose of fear and common sense. Jury is still out on which.

**Comment by scottfurry** [comment]
Pull request, #3424 , created to directly address problem with toolbar. Settings affecting toolbar will be cleaned up with this effort. #3312 can be closed.

**Comment by scottfurry** [comment]
~Staging_Post_TBFix.diff.zip~ removed

Appending snapshot of work-in-progress.

Diff above includes the QToolBar fix from above, as well as, rework of code as a result of other recent commits.

The adventure continues...

**Comment by scottfurry** [comment]
Moving #3341, referencing `sourceforge.net`, to #3419, Code TODO List.
Problem has other implications (app bundle signing) that are beyond settings scope.

**Comment by scottfurry** [comment]
Just to clarify things...The last few posts in this item are intended to address some _Scope Creep Management_. Re-assessing what work is involved to correct issues added appears to be more involved than just settings. Dealing with these items immediately, or deferring to the Code TODO List, #3419, as separate matters seems logical and prudent.

**Comment by scottfurry** [comment]
~Settings_Clean-Up_diff.zip~ (removed - see below)

The zip above contains TWO diffs. Both focus on implementing a single effort. It helps me wrap my head around what I'm doing, as well as preventing straying into _extra_ editing.

First diff is just a straight-up renaming of all the string literals associated with settings storage. The second diff is where some manipulation of the Settings object occurs. 

Everything still builds somewhat cleanly and no bugs appear to have been introduced.

I recently tweaked something in my OS that affects Qt presentation. This has created a glitch with how Qt reports font settings. I have to re-examine font default settings.

Another concern is the syntax highlighting. How this is implemented works but seems a touch ugly to maintain. I'm considering if this functionality can be _somehow_ peeled out and detached from settings. Again, not introducing bugs is key here.

**Reply by justinclift** [reply]
Excellent.  This is looking like really solid progress @scottfurry. :smile:

**Comment by scottfurry** [comment]
~Settings_Clean-Up_diff_2023Aug20.zip~ (removed - see below)

...progress...

A few points...
a) Fixed the issue with a misconfiguration of Fixed Pitch (i.e. monospace or fixed width) fonts. The problem was partly on my end but highlighted a code issue. In `settings.cpp` is code that corrects for possible bad configuration (Qt on Linux is not exactly well integrated). Basically, so long as Qt can find **_ANY_** fixed pitch font, that font will be used for the databrowser, db, editor, and log presentation.

The function `Settings::setDefaultFixedFont` grabs the fixed pitch font reported by `QFontDatabase`. A quick sanity check is done to ensure what is reported is valid (exists in the `QFontDatabase` font family listings **AND** is fixed pitched). If the sanity check fails, _borrowing_ code from Qt's `QFontCombo` control, we dive into font families looking for the first fixed pitch font for the user's locale/writing system (see [`QFontDatabse`](https://doc.qt.io/qt-5/qfontdatabase.html) ). A warning is issued to the user about the situation.

b) The warning...
```
qt.qpa.fonts: Populating font family aliases took 170 ms. Replace uses of missing font family "Monospace" with one that exists to avoid this cost. 
```
still happens when executing on MacOS.

I believe this stems from `[project_root]/libs/qhexedit/src/qhexedit.cpp:31`, assuming building against internal `QHexEdit`. A preprocessor `#ifdef` for platform is employed to set the font in case of Windows platform. The _else_ clause calls out the "Monospace" font explicitly (i.e. non-Windows OS's). This is the only occurrence of "Monospace" in code in the project, as of now. I have to verify that updated dependency corrects this issue (depends on me fixing issues found in #3410).

c) `Settings.cpp` is basically a _re-write_. Previous diffs of this file were practically indecipherable, and why I had to suck back being more pedantic in my approach (step 1, step 2, etc). Git does not recognize a whole-sale change to a file. Best I could do was to tell Git to use a different algorithm for comparison (see [git-diff help - histogram](https://git-scm.com/docs/git-diff#Documentation/git-diff.txt-codehistogramcode) for more details).

d) I'm still not happy about the code for `isValidSettingsFile`. Testing a `QString` passed by reference against `nullptr` still _seems wrong_ to me. I don't have an answer for this one, yet.

There is still a bit of work needed but things seem in a much better place.

... and, of course, I'm finding more Code-ToDo's. Shocking, I know (insert heavy sarcasm).

**Comment by scottfurry** [comment]
~Settings_Clean-Up_diff_2023Aug21.zip~ (removed - see below)

Mechanism to correct for deprecated/bad settings values implemented. On initialization of Settings Object, settings file storage is scanned for section/key pairs that are deprecated or invalid. If item is found the replacement section/key pair is established with found value and old section/key pair is removed. (see `Settings.cpp:processDeprecatedSettings()`)

For now, only one item is being searched. The key `lastlocations`, plural of `lastlocation` used in other places, appears in the src/FileDialog.cpp file.

Also, mechanism exists to remove section/keys no longer in use, i.e. ensure unused settings keys do not remain resident, (see `Settings.cpp:processDeletedSettings()`). At this time, no items are identified but can be added as needed.

Both mechanisms are flexible enough for future changes. The old section/key pair, and replacement for deprecated only, just need to be appended to the appropriate container. Code takes care of the reset.

So what's left?
- I want to take a swipe at `isValidSettingsFile`. I have a notion of how to address this.
- There's some debugging code I have to remove but I would like to keep a couple of functions around for future use. (see `Settings.cpp:debug_[cache|default]`)
- ensure I'm correctly setting syntax highlighting
... and ...
I think that's it. Hopefully the light at the end of tunnel isn't an oncoming train.

**Reply by justinclift** [reply]
Cool, solid progress happening. :smile:

**Comment by scottfurry** [comment]
[Settings_Clean-Up_diffs_2023Sep18.zip](https://github.com/sqlitebrowser/sqlitebrowser/files/12645317/Settings_Clean-Up_diffs_2023Sep18.zip)

And movement.

I was able to crawl through syntax highlighting. There was clean-up. Stuff is _better_. Code feels more solid and, as a bonus, I'm not introducing stuff that shouldn't be there.

Last item on my _TODO_ list is taking another swipe at `isValidSettingsFile`.
Working towards to get this in shape for next release.

Stand By to Stand By.

Edit: Forgot to mention... Had to rebase everything to latest commits. There was a change someone made to set max row count. I missed that commit but was able to incorporate that change into this effort.

**Comment by scottfurry** [comment]
The weeds... A machete was involved...I just barely made it out...

Having reworked a few things, I was enthusiastically doing my due diligence ensuring everything was working _just fine_. Then I got sucked into the vortex that is Qt's QSettings API. It wasn't pretty.

So, what I discovered is that Qt considers the section name "General" to be a reserved name. Any attempt to change will not work. QSettings will automatically rename the section as "%General" and save the result. Of course, the whole matter is _undocumented_. Shocking, I know \<insert heavy sarcasm here\>.

I can sort of see _why_ from Qt's perspective. At the same time, I can see how DB4S code has been configured to work with Qt's imposition. The overwhelming urge to re-write things was enormous. I was seriously questioning the whole validity of the exercise thinking I was misunderstanding the read/write of settings. Again, ugly.

I have some changes to undo. And a bug report to file (Damn it. Document this quirk Qt!).
Suck back. Reload. Try Again. Soon...

Edit: FTR, [QT Bug 117427](https://bugreports.qt.io/browse/QTBUG-117427)

**Comment by scottfurry** [comment]
Pull Request #3481 - It is done...

**Comment by scottfurry** [comment]
### Overview
Let's wrap this up with a quick discussion about code changes and impact.

#### Start Up and Settings Initialization
On App start the environment variable `DB4S_SETTINGS_FILE` is checked for a value, doing away with the `!= nullptr` expressions, and if empty we check the command line for the `-S` switch. If string value is found, it is verified as pointing to a valid settings file.

The settings object is started. The settings are then scanned to remove deleted settings values (`m_hDeleted` settings variable). Any deprecated settings found are replaced with new values (`m_hDeprecated` settings variable). Finally, the value for font undergoes a _sanity check_, and if necessary, replaced with first fixed font from system. Finally, the settings for default colours, based on selected theme setting, are established.

#### Get Value
This mechanism is reworked a bit. The order of preference is still to retrieve the cache value before accessing the settings file. Added is the mechanism that if neither contains the desired section/key value, the application falls back to a default value (`m_hDefault` settings variable). Defaults are established in source file. If the section/key is not in any of cache, file, or defaults a warning is issued on the console.

#### Going Forward
Settings values will need to be managed in future. INI file section/keys are explicitly defined as `const std::string` constants to prevent misspelled entries from causing problems. This is what happened with the setting for `lastlocation` (a plural version existed elsewhere in code - corrected in deprecated settings handling).

New Section/Keys will need to be added to the `src/Settings.h` file under the `szINI` namespace. The default value for any Key will need to be added to the top of the `src/Settings.cpp` file where the `m_hDefault` unordered_map is defined.

If the new key involves fonts or colours, entry default should be defined as an empty QVariant and then populated in the `processFixedFont` or `processDefaultColours` function.

#### Final
a) QHexEdit is causing noise about the "Monospace" font. This stems from the current version of code having an if statement that explicitly sets the control font to "Monospace" for non-Windows platforms. I believe this issue is corrected upstream in later versions. For MacOS platforms, this warning can be ignored as the font for the control is set via fixed font default functionality on application start.

b) In setting the fixed font, functions from the QFontCombo control were pulled into DB4S source. I find it incredible that useful methods to programmatically determine system fixed font is buried in a GUI control. [QTBUG 116223](https://bugreports.qt.io/browse/QTBUG-116223) is a feature request to QT to move the functions `writingSystemFromScript` and `writingSystemFromLocale` from the GUI control into `QFontDatabase`. If implemented, these two functions, pulled from Qt Sources, can be removed from `src/Settings.[h|cpp]` files.
